File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        left, right = 0, len(matrix[0]) - 1
        top, bottom = 0, len(matrix) - 1
        result = []

        while left < right and top < bottom:
            for col in range(left, right):
                result.append(matrix[top][col])

            for row in range(top, bottom):
                result.append(matrix[row][right])

            for col in range(right, left, -1):
                result.append(matrix[bottom][col])

            for row in range(bottom, top, -1):
                result.append(matrix[row][left])

            left += 1
            right -= 1
            top += 1
            bottom -= 1

        for row in range(top, bottom + 1):
            for col in range(left, right + 1):
                result.append(matrix[row][col])

        return result

File: 0054-spiral-matrix/0054-spiral-matrix.js
const spiralOrder = (matrix) => {
  const result = [];
  
  const [rowsSize, colsSize] = [matrix.length - 1, matrix[0].length - 1];
  let [left, right, top, bottom] = [0, colsSize, 0, rowsSize];

  const isInBounds = () => ((left <= right) && (top <= bottom));

  while (isInBounds()) {
    addTop({ matrix, result, left, right, top });
    top++;

    addRight({ matrix, result, right, top, bottom });
    right--;

    if (!isInBounds()) break;

    addBottom({ matrix, result, left, right, bottom });
    bottom--;

    addLeft({ matrix, result, left, top, bottom });
    left++;
  }

  return result;
};

const addTop = ({ matrix, result, left, right, top }) => {
  for (let col = left; col <= right; col++) {
    result.push(matrix[top][col]);
  }
};

const addRight = ({ matrix, result, right, top, bottom }) => {
  for (let row = top; row <= bottom; row++) {
    result.push(matrix[row][right]);
  }
};

const addBottom = ({ matrix, result, left, right, bottom }) => {
  for (let col = right; col >= left; col--) {
    result.push(matrix[bottom][col]);
  }
};

const addLeft = ({ matrix, result, left, top, bottom }) => {
  for (let row = bottom; row >= top; row--) {
    result.push(matrix[row][left]);
  }
};
